### PR TITLE
[B2BP-503] - Terraform CDN for multitenancy

### PR DESCRIPTION
**Pull request by maurodandrea**
**PR in draft until the SSL certified validated by the SEND group**

Create with Terraform multiple CDN for multitenancy with "for_each" statement.
PS: for now, the existing CDN "aws_cloudfront_distribution" "website" it has not been removed and new CDN is created with default cloudfront certificate

Depends on #237
Depends on #247

#### List of Changes
variable.tf and cloudfront.tf terraform file.

#### Motivation and Context
Create with Terraform multiple CDN for multitenancy with "for_each" statement.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [X] My change not requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/infrastructure/src/cloudfront.tf [hunk 1]
```diff
@@ -90,12 +90,79 @@ resource "aws_cloudfront_distribution" "website" {
...
+      restriction_type = "none"
+    }
+  }
+
+  viewer_certificate {
+    # set default = true in variable "use_custom_certificate" when available and validate a tenant certificate
+    cloudfront_default_certificate = var.use_custom_certificate ? false : true
+    acm_certificate_arn            = var.use_custom_certificate ? module.cdn_websites_ssl_certificate[each.key].acm_certificate_arn : null
+    ssl_support_method             = var.use_custom_certificate ? "sni-only" : null
+  }
```
Comment (kin0992) [inline apps/infrastructure/src/cloudfront.tf:163]: **issue (blocking)**: add `minimum_protocol_version`

here you should add the option `minimum_protocol_version = "TLSv1.2_2021"`, due to an internal requirement [described in this page](https://pagopa.atlassian.net/wiki/spaces/DEVOPS/pages/965641850/AWS+Production+Security+checklist) (but I'm not sure you have access to it).

File: apps/infrastructure/src/cloudfront.tf [hunk 2]
```diff
@@ -90,12 +90,79 @@ resource "aws_cloudfront_distribution" "website" {
...
+    response_code      = 404
+    response_page_path = "/404.html"
+  }
+
+  default_cache_behavior {
+    # HTTPS requests we permit the distribution to serve
+    allowed_methods  = ["GET", "HEAD", "OPTIONS"]
+    cached_methods   = ["GET", "HEAD"]
+    target_origin_id = aws_s3_bucket.website.bucket
+    #response_headers_policy_id = aws_cloudfront_response_headers_policy.websites.id
```
Comment (kin0992) [inline apps/infrastructure/src/cloudfront.tf:130]:
```suggestion
```

File: apps/infrastructure/src/cloudfront.tf [hunk 3]
```diff
@@ -90,12 +90,79 @@ resource "aws_cloudfront_distribution" "website" {
...
+    }
+    origin_path = each.value.origin_path
+  }
+
+  enabled             = true # enable CloudFront distribution
+  is_ipv6_enabled     = true
+  comment             = "CloudFront distribution for the static website."
+  default_root_object = "index.html"
+
+  # to be uncommented when available dns domain and certificate is validated
```
Comment (kin0992) [inline apps/infrastructure/src/cloudfront.tf:116]: **question**: what should be uncommented?

**Review by kin0992 — COMMENTED** [review]
you should also add the changeset

File: apps/infrastructure/src/cloudfront.tf (hunk 1 above)
Comment (maurodandrea) [reply apps/infrastructure/src/cloudfront.tf:162]: added in next commit

File: apps/infrastructure/src/cloudfront.tf (hunk 3 above)
Comment (maurodandrea) [reply apps/infrastructure/src/cloudfront.tf:116]: comment removed in next commit

File: .changeset/dirty-houses-run.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"infrastructure": minor
+---
+
+[B2BP-503] - Terraform CDN for multitenancy
```
Comment (kin0992) [inline .changeset/dirty-houses-run.md:5]: @maurodandrea what you write here will be visible into [the CHANGELOG file](https://github.com/pagopa/b2b-portals/blob/b06ec3a6cb23a7a96d977de7505af47831e1539b/apps/infrastructure/CHANGELOG.md).

so, my question is: do you think that message is valuable to someone who will read the changelog?

File: apps/infrastructure/src/variables.tf [hunk 5]
```diff
@@ -50,7 +50,7 @@ variable "cms_app_memory" {
 variable "use_custom_certificate" {
   type        = bool
   description = "Enable CDN https support with a custom certificate instead using the default one"
-  default     = false # set true when available dns and custom certificate
+  default     = true # set true when available dns and custom certificate
```
Comment (kin0992) [inline apps/infrastructure/src/variables.tf:53]: **question (non-blocking)**: is the comment needed? or is the change needed?

I think the comment and the change are discordant

File: apps/infrastructure/src/cloudfront.tf [hunk 6]
```diff
@@ -90,12 +90,78 @@ resource "aws_cloudfront_distribution" "website" {
...
+    origin_id   = aws_s3_bucket.website.bucket
+    s3_origin_config {
+      origin_access_identity = aws_cloudfront_origin_access_identity.main.cloudfront_access_identity_path
+    }
+    origin_path = each.value.origin_path
+  }
+
+  enabled             = true # enable CloudFront distribution
+  is_ipv6_enabled     = true
+  comment             = "CloudFront distribution for the static website."
```
Comment (christian-calabrese) [inline apps/infrastructure/src/cloudfront.tf:113]: Nel commento della cloudfront distribution inserirei il nome del sito per chiarire a quale tenant si riferisce. Un idea potrebbe essere concatenare la chiave each.key

Comment (maurodandrea) [reply apps/infrastructure/src/cloudfront.tf:113]: ok, i've set the comment = "CloudFront distribution for the static website ${each.key}"

File: apps/infrastructure/src/variables.tf (hunk 5 above)
Comment (maurodandrea) [reply apps/infrastructure/src/variables.tf:53]: ok, i've change the comment

File: .changeset/dirty-houses-run.md (hunk 4 above)
Comment (maurodandrea) [reply .changeset/dirty-houses-run.md:5]: ok, i've updated the message

**Pull request merged**